File: cpp/165.cpp
#include <iostream>

#define REP(i,a) for(int i=0;i<(a);i++)

using namespace std;
typedef long long ll;

//Note: this program is broken since there are some duplicate intersections.
//see 165.py
//One speedup would be to use this code to calculate intersecting segments,
//then use python to calculate intersections and remove duplicates.
//Or I could not be lazy and just do it in C.

const int mod=50515093;
const int N=5000;
int p[N][4];

inline int sgn(int x) {
    return (x>0)-(x<0);
}

inline bool intersects(int i, int j) {
    int xi1=p[i][0],yi1=p[i][1],xi2=p[i][2],yi2=p[i][3];
    int xj1=p[j][0],yj1=p[j][1],xj2=p[j][2],yj2=p[j][3];
    //straddle each other
    if (sgn(sgn((xj1-xi1)*(yi2-yi1)-(xi2-xi1)*(yj1-yi1)) * sgn((xj2-xi1)*(yi2-yi1)-(xi2-xi1)*(yj2-yi1))) >= 0) return false;
    //straddle each other
    if (sgn(sgn((xi1-xj1)*(yj2-yj1)-(xj2-xj1)*(yi1-yj1)) * sgn((xi2-xj1)*(yj2-yj1)-(xj2-xj1)*(yi2-yj1))) >= 0) return false;
    return true;
}

int main() {
    int total=0;
    ll s=290797;
    REP(i,N) {
        REP(j,4) {
            s=s*s%mod;
            p[i][j]=s%500;
        }
    }
    REP(i,N) REP(j,i) if (intersects(i,j)) {
        total++;
    }
    cout << total << endl;
    return 0;
}

File: python/165.py
#!/usr/bin/env python
from gmpy import mpq

mod=50515093
N=5000
p=[]
pts=set([])

def sgn(x):
    return (x>0)-(x<0)

def intersects(i, j):
    xi1,yi1,xi2,yi2=p[i][0],p[i][1],p[i][2],p[i][3]
    xj1,yj1,xj2,yj2=p[j][0],p[j][1],p[j][2],p[j][3]
    if sgn(sgn((xj1-xi1)*(yi2-yi1)-(xi2-xi1)*(yj1-yi1)) * sgn((xj2-xi1)*(yi2-yi1)-(xi2-xi1)*(yj2-yi1))) >= 0: return False
    if sgn(sgn((xi1-xj1)*(yj2-yj1)-(xj2-xj1)*(yi1-yj1)) * sgn((xi2-xj1)*(yj2-yj1)-(xj2-xj1)*(yi2-yj1))) >= 0: return False
    xp=None
    if xi2==xi1: xp=xi1
    else: mi=mpq(yi2-yi1)/mpq(xi2-xi1)
    if xj2==xj1: xp=xj1
    else: mj=mpq(yj2-yj1)/mpq(xj2-xj1)
    if xp==None: xp=mpq(yj1-yi1+mi*xi1-mj*xj1)/mpq(mi-mj)
    if xi1!=xi2: yp=mi*(xp-xi1)+yi1
    else: yp=mj*(xp-xj1)+yj1
    return (xp,yp)

total=0;
s=290797;
for i in xrange(N):
    t=[]
    for j in xrange(4):
        s=s*s%mod
        t.append(s%500)
    p.append(t)

for i in xrange(N):
    #print i
    for j in xrange(i):
        t=intersects(i,j)
        if t: pts.add(t)
print len(pts)
